_posts/2022-03-02-gcp-workload-identity-with-gke-and-terraform.markdown

```diff
@@ -0,0 +1,287 @@
+```sh
+#!/usr/bin/env bash
+
+set -ex -o pipefail -o nounset
+
+K8S_NAMESPACE="cloudlad"
+K8S_SERVICE_ACCOUNT="cloudlad"
+GCP_REGION="europe-west3"
+CLUSTER_NAME="cloudlad-2-ringtail"
+GCP_PROJECT_ID="workload-identity-2-ringtail"
+K8S_CONTEXT="gke_workload-identity-2-ringtail_europe-west3_cloudlad-2-ringtail"
+GCP_SERVICE_ACCOUNT="cloudlad-2-ringtail-k8s@workload-identity-2-ringtail.iam.gserviceaccount.com"
+
+gcloud container clusters get-credentials "$CLUSTER_NAME" --project="$GCP_PROJECT_ID" --region="$GCP_REGION"
+
+kubectl config use-context "$K8S_CONTEXT"
+
+kubectl create namespace "$K8S_NAMESPACE"
+
+kubectl create serviceaccount "$K8S_SERVICE_ACCOUNT" \
+    --namespace="$K8S_NAMESPACE"
+
+kubectl annotate serviceaccount "$K8S_SERVICE_ACCOUNT" \
+    --namespace="$K8S_NAMESPACE" \
+    iam.gke.io/gcp-service-account="$GCP_SERVICE_ACCOUNT"
+
+kubectl run --image=google/cloud-sdk:slim \
+    --namespace="$K8S_NAMESPACE" \
+    --overrides="{ 'spec': { 'serviceAccount': $K8S_SERVICE_ACCOUNT } }" \
+    cloudlad -- sleep infinity
+```
+
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 6. Helper shell script for programmatically creating the K8s resources, valid only for a single Terraform state.</b>
+</p>
+
+When I execute the k8s.sh script, the K8s Service Account will get annotated with a key: iam.gke.io/gcp-service-account and a value: cloudlad-2-ringtail-k8s@workload-identity-2-ringtail.iam.gserviceaccount.com. The annotation enables the K8s service account to inherit the same Authorization scope from the IAM Service account. See Figure 7.
+
+```json
+$ kubectl -n cloudlad get serviceaccounts cloudlad -o json
+{
+    "apiVersion": "v1",
+    "kind": "ServiceAccount",
+    "metadata": {
+        "annotations": {
+            "iam.gke.io/gcp-service-account": "cloudlad-2-ringtail-k8s@workload-identity-2-ringtail.iam.gserviceaccount.com"
+        },
+        "creationTimestamp": "2022-02-22T21:28:59Z",
+        "name": "cloudlad",
+        "namespace": "cloudlad",
+        "resourceVersion": "9849",
+        "uid": "f29f981f-28fc-4ca2-ba35-31dd42ff8e7a"
+    },
+    "secrets": [
+        {
+            "name": "cloudlad-token-thlzr"
+        }
+    ]
+}
+```
+<p align="center">
+  <b style="font-size:0.7vw" >Figure 7. Service Account annotation got with kubectl.</b>
+</p>
+
+You might ask yourself, "OK, What now?". Checking whether the coupling between GCP IAM and GKE works is straightforward. The newly created pod runs indefinite sleep. And the container is started from the official GCP SDK image, meaning the gcloud CLI is readily available.
+
+I will assume a shell in the running container. Then check the available service accounts, and finally, create a PubSub topic. It is all possible because the pod inherits the [roles/pubsub.admin] role. See Figure 8.1.
```

Comment:
```suggestion
I will assume a shell is in the running container. Then check the available service accounts, and finally, create a PubSub topic. It is all possible because the pod inherits the [roles/pubsub.admin] role. See Figure 8.1.
```

Reply:
I will change the wording a bit here. By assume I meant execute a shell.